control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs

```diff
@@ -143,21 +159,66 @@ async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> Po
-            .specs()
-            .set_node_drained(context.registry(), node_spec.id())
-            .await
-        {
-            tracing::error!(
-                %error,
-                node.id = node_id.as_str(),
-                "Failed to set node to state drained"
-            );
-            return PollResult::Err(error);
+        // All volumes on the node are republished.
+        // Determine whether we can mark the node as drained.
+        // If not, the next reconciliation loop will check again.
+        let mut can_set_drained = true;
+
+        let drainingvolumes = context.specs().get_node_draining_volumes(node_id).await?;
+        for (vi, starttime) in drainingvolumes {
+            let elapsed = starttime.elapsed();
+            let mut remove_vol_from_list = true;
+            if elapsed.is_ok()
+                && elapsed.unwrap() < Duration::from_secs(DRAINING_VOLUME_TIMOUT_SECONDS)
+                && context.specs().volume_rsc(&vi).is_some()
+            {
+                // if the drain is within the timeout and the volume still exists
+                let shutdown_nexuses = context
+                    .registry()
+                    .specs()
+                    .volume_shutdown_nexuses(&vi)
+                    .await;
+                if !shutdown_nexuses.is_empty() {
+                    // if it still has shutdown nexuses
+                    can_set_drained = false; // don't set node to drained
+                    remove_vol_from_list = false; // keep it in the list for the next iteration
+                    tracing::info!(
+                        nexus.count = shutdown_nexuses.len(),
+                        "Shutdown nexuses remain"
+                    );
+                } // else it is no longer shutdown
+            } // else it is timed out or no longer exists
+            if remove_vol_from_list {
+                tracing::info!(
+                    node.id = node_id.as_str(),
+                    volume.id = vi.as_str(),
+                    "Removing a volume from the draining volume list"
+                );
+                context
+                    .specs()
+                    .remove_node_draining_volume(context.registry(), node_id, &vi)
+                    .await?;
+            }
+        }
+        // Change the node state to "drained"
+        if can_set_drained {
+            if let Err(error) = context
+                .specs()
+                .set_node_drained(context.registry(), node_id)
+                .await
+            {
+                tracing::error!(
+                    %error,
+                    node.id = node_id.as_str(),
+                    "Failed to set node to state drained"
+                );
+                return PollResult::Err(error);
+            }
+            tracing::info!(node.id = node_id.as_str(), "Set node to state drained");
+        } else {
+            tracing::info!(node.id = node_id.as_str(), "Shutdown nexuses remain");
```

Comment:
Can we keep a count?
EDIT: made suggestion above

Reply:
fixed